js/packages/teams-ai/src/authentication/Authentication.ts

```diff
@@ -0,0 +1,110 @@
+        botAuth?: BotAuthentication<TState>
+    ) {
+        this._oauthPromptSettings = settings;
+        this._messagingExtensionAuth = messagingExtensionsAuth || new MessagingExtensionAuthentication(settings);
+        this._botAuth = botAuth || new BotAuthentication(app, settings, storage);
+    }
+
+    /**
+     * Signs in a user.
+     *
+     * This method will be called automatically by the Application class.
+     * @template TState
+     * @param {TurnContext} context - Current turn context.
+     * @param {TState} state Application state.
+     * @returns {string | undefined} The authentication token or undefined if the user is still login in.
+     */
+    public async signInUser(context: TurnContext, state: TState): Promise<string | undefined> {
+        if (this.isMessagingExtensionAuthFlow(context)) {
+            return await this._messagingExtensionAuth.authenticate(context, state);
+        } else if (this.isBotAuthFlow(context)) {
+            return await this._botAuth.authenticate(context, state);
+        } else {
+            throw new Error(`signInUser() is not supported for this activity type.`);
+        }
+    }
+
+    /**
+     * Signs out a user.
+     * @template TState
+     * @param {TurnContext} context - Current turn context.
+     * @param {TState} state - Application state.
+     * @returns {Promise<void>} A Promise representing the asynchronous operation.
+     */
+    public signOutUser(context: TurnContext, state: TState): Promise<void> {
+        if (this.isBotAuthFlow(context)) {
+            this._botAuth.deleteAuthFlowState(context, state);
+        }
+
+        // Signout flow is agnostic of the activity type.
+        return UserTokenAccess.signOutUser(context, this._oauthPromptSettings);
+    }
+
+    /**
+     * Determines whether user sign in is allowed for the incomming activity type.
+     * @param {TurnContext} context - Current turn context.
+     * @returns {boolean} true if the user can sign in for the incomming activity type, false otherwise.
+     */
+    public canSignInUser(context: TurnContext): boolean {
+        return this.isBotAuthFlow(context) || this.isMessagingExtensionAuthFlow(context);
+    }
+
+    /**
+     * The handler function is called when the user has successfully signed in.
+     *
+     * This only applies if sign in was initiated by the user sending a message to the bot.
+     * This handler will not be triggered if a messaging extension triggered the authentication flow.
+     * @template TState
+     * @param {(context: TurnContext, state: TState) => Promise<void>} handler The handler function to call when the user has successfully signed in
+     */
+    public async onUserSignIn(handler: (context: TurnContext, state: TState) => Promise<void>): Promise<void> {
+        this._botAuth.onUserSignIn(handler);
+    }
+
+    private isBotAuthFlow(context: TurnContext): boolean {
+        return context.activity.type == ActivityTypes.Message;
```

Comment:
The Adaptive Cards submit event is also a message activity: https://github.com/microsoft/teams-ai/blob/main/js/packages/teams-ai/src/AdaptiveCards.ts#L346
I don't know whether there are more cases, may need to update the constraint here.

Reply:
Good catch. Added check to make sure the `context.activity.text` is populated with a non-empty string.